lilcommodule.c

```diff
@@ -1,39 +1,307 @@
+    output = malloc(sizeof(int8_t) * n_samples * output_stride);
+    
+
+    /* Conversion to int16_t array */
+    switch (((PyArrayObject*)signal_object)->descr->type_num){
+      case NPY_INT8:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((int8_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_INT16:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((int16_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_INT32:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((int32_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_INT64:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((int64_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_UINT8:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((uint8_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_UINT16:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((uint16_t *)signal_object_data)[i]); 
+        } break;
+      case NPY_UINT32:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((uint32_t *)signal_object_data)[i]);
+        } break;
+      case NPY_UINT64:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (int16_t)(((uint64_t *)signal_object_data)[i]); 
+        } break;
+    }
+
+    /* Calling the core function */
+    lilcom_compress(n_samples, input, input_stride, output, output_stride, lpc_order);
+
+    /* Debug: Comment or Uncomment when on debug */
+    // for (int i = 0; i < n_samples * input_stride ; i++){
+    //   printf("for index %d a = %d and b = %d\n", i , input[i], output[i]);
+    // }
+
+    /* Making the resulting array */
+    npy_intp * output_dimensions = malloc(sizeof(npy_intp)*2);
+    output_dimensions[0] = n_samples;
+    output_dimensions[1] = output_stride;
+    PyArrayObject * output_array = (PyArrayObject *) PyArray_SimpleNewFromData(n_dims, output_dimensions, NPY_INT8, (void*) output);
+
+    /* Overcoming memory leak problem */
+    free(input);
+    /* Returning numpy array */
+    PyObject *returner = PyArray_Return(output_array);
+    return returner;
+  } 
+  
+  else { // Float
+    float *input; // The one dimensional vectorized input array which will be given to the core function
+    /* Allocating the space for input and output */
+    input = malloc(sizeof(float) * n_samples * input_stride);
+    output = malloc(sizeof(int8_t) * n_samples * output_stride);
+    
+
+    /* Conversion to int16_t array */
+    switch (((PyArrayObject*)signal_object)->descr->type_num){
+      case NPY_FLOAT:
+        for (int i = 0 ; i < n_samples * input_stride ; i++){
+          input[i] = (float)(((float *)signal_object_data)[i]); 
```

Comment:
This isn't quite right, we need to think about the scale of the signal.  Normally in Python, signals represented as floating point will be  in the range [0, 1], so we'd have to scale by 32768.0 before casting to int16.  But I don't want to assume that.  There's some stuff I want to do at the Python level anyway as it's easier.  For now just focus on getting the very simple case where the input is int16_t working.

Reply:
I actually wasn't in the context of how to manipulate the float and I thought it would be done in the core library. So we can also delete the case where the input might be float inside the extension module. Do you agree?